spec/integration/client_options_spec.rb

```diff
@@ -0,0 +1,441 @@
+
+        it 'correctly sets client options' do
+          expect(client.options[:auth_mech_properties]).to eq({ 'service_name' => 'mongodb' })
+        end
+      end
+    end
+
+    # TODO: get this test passing
+    # context 'with client options' do
+    #   let(:client_opts) do
+    #     {
+    #       auth_mech: :gssapi,
+    #       user: user,
+    #       password: pwd
+    #     }
+    #   end
+
+    #   it 'sets default auth mech properties' do
+    #     expect(client.options[:auth_mech_properties]).to eq({ 'service_name' => 'mongodb' })
+    #   end
+    # end
+  end
+
+  context 'with PLAIN auth mechanism' do
+    let(:auth_mech_string) { 'PLAIN' }
+    let(:auth_mech_sym) { :plain }
+
+    it_behaves_like 'a supported auth mechanism'
+    it_behaves_like 'auth mechanism that uses database or default auth source', default_auth_source: '$external'
+    it_behaves_like 'an auth mechanism with ssl'
+    it_behaves_like 'an auth mechanism that doesn\'t support auth_mech_properties'
+  end
+
+  context 'with MONGODB-X509 auth mechanism' do
+    let(:auth_mech_string) { 'MONGODB-X509' }
+    let(:auth_mech_sym) { :mongodb_x509 }
+
+    let(:pwd) { nil }
+
+    it_behaves_like 'a supported auth mechanism'
+    it_behaves_like 'an auth mechanism with ssl'
+    it_behaves_like 'an auth mechanism that doesn\'t support auth_mech_properties'
+    it_behaves_like 'an auth mechanism that doesn\'t support invalid auth sources'
+
+    context 'with URI options' do
+      let(:credentials) { "#{user}@" }
+      let(:options) { '?authMechanism=MONGODB-X509' }
+
+      it 'sets default auth source' do
+        expect(client.options[:auth_source]).to eq('$external')
+      end
+
+      context 'when username is not provided' do
+        let(:credentials) { '' }
+
+        it 'recognizes the mechanism with no username' do
+          expect(client.options[:user]).to be_nil
+        end
+      end
+
+      context 'when a password is provided' do
+        let(:credentials) { "#{user}:password@" }
+
+        it 'throws an error on client creation' do
+          expect {
+            client
+          }.to raise_error(Mongo::Auth::InvalidConfiguration, /password is not supported/)
+        end
+      end
+    end
+
+    context 'with client options' do
+      let(:client_opts) { { auth_mech: :mongodb_x509, user: user } }
+
+      # TODO: get this test passing
+      # it 'sets default auth source' do
+      #   expect(client.options[:auth_source]).to eq('$external')
+      # end
+
+      context 'when username is not provided' do
+        let(:client_opts) { { auth_mech: :mongodb_x509} }
+
+        it 'recognizes the mechanism with no username' do
+          expect(client.options[:user]).to be_nil
+        end
+      end
+
+      context 'when a password is provided' do
+        let(:client_opts) { { auth_mech: :mongodb_x509, user: user, password: 'password' } }
+
+        it 'throws an error on client creation' do
```

Comment:
In Ruby exceptions are raised. There is also a throw keyword which does something else https://stackoverflow.com/questions/51021/what-is-the-difference-between-raising-exceptions-vs-throwing-exceptions-in-ruby).